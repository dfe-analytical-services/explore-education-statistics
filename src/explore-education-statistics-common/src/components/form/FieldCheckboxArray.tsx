import createErrorHelper from '@common/validation/createErrorHelper';
import {
  connect,
  FieldArray,
  FieldArrayRenderProps,
  FormikContext,
} from 'formik';
import get from 'lodash/get';
import React, { ReactNode } from 'react';

interface InjectedProps<FormValues> extends FieldArrayRenderProps {
  error: string;
  value: FormValues[keyof FormValues];
}

interface FieldCheckboxArrayProps<FormValues> {
  children: (props: InjectedProps<FormValues>) => ReactNode;
  error?: string;
  name: string;
  showError?: boolean;
  validateOnChange?: boolean;
}

const FieldCheckboxArray = <T extends {}>({
  children,
  error,
  name,
  formik,
  showError = true,
  validateOnChange = false,
}: FieldCheckboxArrayProps<T> & { formik: FormikContext<T> }) => {
  return (
    <FieldArray
      name={name}
      validateOnChange={validateOnChange || formik.validateOnChange}
    >
      {({ form, ...arrayHelpers }) => {
        const { getError } = createErrorHelper(form);

        let errorMessage = error || getError(name);

        if (!showError) {
          errorMessage = '';
        }

        const value = get(form.values, name);

        return children({
          ...arrayHelpers,
          error: errorMessage,
          form,
          name,
          value,
        });
      }}
    </FieldArray>
  );
};

export default connect<FieldCheckboxArrayProps<{}>>(FieldCheckboxArray);
