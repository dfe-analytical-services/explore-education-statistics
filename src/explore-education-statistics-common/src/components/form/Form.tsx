import ErrorSummary, {
  ErrorSummaryMessage,
} from '@common/components/ErrorSummary';
import { FormIdContextProvider } from '@common/components/form/contexts/FormIdContext';
import useMountedRef from '@common/hooks/useMountedRef';
import useToggle from '@common/hooks/useToggle';
import logger from '@common/services/logger';
import isErrorLike from '@common/utils/error/isErrorLike';
import createErrorHelper from '@common/validation/createErrorHelper';
import { useFormikContext } from 'formik';
import camelCase from 'lodash/camelCase';
import get from 'lodash/get';
import isEqual from 'lodash/isEqual';
import React, {
  ReactNode,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

interface Props {
  children: ReactNode;
  id: string;
  submitId?: string;
  showErrorSummary?: boolean;
  showSubmitError?: boolean;
}

/**
 * Form wrapper to integrate with Formik.
 *
 * This provides a bunch of conveniences for displaying errors
 * and linking to them correctly (in error message hrefs) as
 * long as certain conventions are followed.
 *
 * Fields with errors should have ids which share the form's
 * id and camelCased value key in Formik
 * e.g. if form id is `timePeriodForm`, then a Formik value with a
 * key of `startDate` should have a field with an id of
 * `timePeriodForm-startDate`.
 *
 * Additional errors upon submitting the form e.g. from server
 * requests will also be added to the error summary. These
 * will link to the `submitId` prop.
 */
const Form = ({
  children,
  id,
  submitId = `${id}-submit`,
  showErrorSummary = true,
  showSubmitError = false,
}: Props) => {
  const isMounted = useMountedRef();

  const formik = useFormikContext();
  const { errors, touched, values, submitCount, submitForm } = formik;

  const previousValues = useRef(values);

  const { getAllErrors } = createErrorHelper({
    errors,
    touched,
  });

  const [hasSummaryFocus, toggleSummaryFocus] = useToggle(false);
  const [submitError, setSubmitError] = useState<ErrorSummaryMessage>();

  useEffect(() => {
    if (!isMounted.current) {
      return;
    }

    // If form has changed at all, we should remove the submit error
    if (!submitCount || !isEqual(values, previousValues.current)) {
      setSubmitError(undefined);
    }

    previousValues.current = values;
  }, [submitError, submitCount, values, isMounted]);

  const allErrors = useMemo(() => {
    const summaryErrors: ErrorSummaryMessage[] = Object.entries(getAllErrors())
      .filter(([errorName]) => get(touched, errorName))
      .map(([errorName, message]) => ({
        id: `${id}-${camelCase(errorName)}`,
        message: typeof message === 'string' ? message : '',
      }));

    return submitError ? [...summaryErrors, submitError] : summaryErrors;
  }, [getAllErrors, id, submitError, touched]);

  const handleSubmit = useCallback(
    async event => {
      toggleSummaryFocus.off();
      setSubmitError(undefined);
      event.preventDefault();

      try {
        await submitForm();
      } catch (error) {
        if (showSubmitError && isErrorLike(error)) {
          logger.error(error);

          if (isMounted.current) {
            setSubmitError({
              id: submitId,
              message: error.message,
            });
          }
        } else {
          throw error;
        }
      } finally {
        if (isMounted.current) {
          toggleSummaryFocus.on();
        }
      }
    },
    [toggleSummaryFocus, submitForm, showSubmitError, isMounted, submitId],
  );

  return (
    <FormIdContextProvider id={id}>
      <form id={id} onSubmit={handleSubmit}>
        {showErrorSummary && (
          <ErrorSummary
            errors={allErrors}
            id={`${id}-summary`}
            focusOnError={hasSummaryFocus}
            onFocus={toggleSummaryFocus.off}
          />
        )}

        {children}
      </form>
    </FormIdContextProvider>
  );
};

export default Form;
