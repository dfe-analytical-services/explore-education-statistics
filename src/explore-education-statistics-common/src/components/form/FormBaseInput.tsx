import ErrorMessage from '@common/components/ErrorMessage';
import FormLabel, { FormLabelProps } from '@common/components/form/FormLabel';
import classNames from 'classnames';
import React, {
  ChangeEventHandler,
  FocusEventHandler,
  KeyboardEventHandler,
  memo,
  MouseEventHandler,
  ReactNode,
} from 'react';

export interface FormBaseInputProps extends FormLabelProps {
  addOn?: ReactNode;
  addOnContainerClassName?: string;
  className?: string;
  disabled?: boolean;
  error?: ReactNode | string;
  hint?: string;
  id: string;
  list?: string;
  name: string;
  width?: 20 | 10 | 5 | 4 | 3 | 2;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  onClick?: MouseEventHandler<HTMLInputElement>;
  onKeyPress?: KeyboardEventHandler<HTMLInputElement>;
}

interface HiddenProps {
  defaultValue?: string | number;
  type?: 'text' | 'number' | 'color' | 'search';
  value?: string | number;
}

const FormBaseInput = ({
  addOn,
  addOnContainerClassName,
  className,
  error,
  hint,
  id,
  hideLabel,
  label,
  labelSize,
  width,
  type = 'text',
  ...props
}: FormBaseInputProps & HiddenProps) => {
  const input = (
    <input
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      aria-describedby={
        classNames({
          [`${id}-error`]: !!error,
          [`${id}-hint`]: !!hint,
        }) || undefined
      }
      className={classNames('govuk-input', className, {
        [`govuk-input--width-${width}`]: width !== undefined,
        'govuk-input--error': !!error,
      })}
      id={id}
      type={type}
    />
  );

  return (
    <>
      <FormLabel
        id={id}
        label={label}
        labelSize={labelSize}
        hideLabel={hideLabel}
      />

      {hint && (
        <span id={`${id}-hint`} className="govuk-hint">
          {hint}
        </span>
      )}

      {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}
      {addOn ? (
        <div className={classNames('dfe-flex', addOnContainerClassName)}>
          {input}
          {addOn}
        </div>
      ) : (
        input
      )}
    </>
  );
};

export default memo(FormBaseInput);
