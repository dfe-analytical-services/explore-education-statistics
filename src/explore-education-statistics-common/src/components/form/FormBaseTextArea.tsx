import ErrorMessage from '@common/components/ErrorMessage';
import FormLabel, { FormLabelProps } from '@common/components/form/FormLabel';
import classNames from 'classnames';
import React, {
  ChangeEventHandler,
  FocusEventHandler,
  KeyboardEventHandler,
  MouseEventHandler,
  ReactNode,
  Ref,
  memo,
} from 'react';

export interface FormTextAreaProps
  extends Pick<FormLabelProps, 'hideLabel' | 'label'> {
  className?: string;
  disabled?: boolean;
  error?: ReactNode | string;
  hint?: string;
  id: string;
  inputRef?: Ref<HTMLTextAreaElement>;
  maxLength?: number;
  name: string;
  rows?: number;
  value?: string;
  onBlur?: FocusEventHandler<HTMLTextAreaElement>;
  onChange?: ChangeEventHandler<HTMLTextAreaElement>;
  onClick?: MouseEventHandler<HTMLTextAreaElement>;
  onKeyPress?: KeyboardEventHandler<HTMLTextAreaElement>;
}

function FormBaseTextArea({
  className,
  disabled,
  error,
  hint,
  id,
  inputRef,
  hideLabel,
  label,
  maxLength,
  name,
  rows = 5,
  value,
  onBlur,
  onChange,
  onClick,
  onKeyPress,
}: FormTextAreaProps) {
  return (
    <>
      <FormLabel id={id} label={label} hideLabel={hideLabel} />

      {hint && (
        <div id={`${id}-hint`} className="govuk-hint">
          {hint}
        </div>
      )}

      {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}

      <textarea
        aria-describedby={
          classNames({
            [`${id}-error`]: !!error,
            [`${id}-hint`]: !!hint,
            [`${id}-info`]: !!maxLength,
          }) || undefined
        }
        className={classNames('govuk-textarea', className, {
          'govuk-textarea--error': !!error,
        })}
        disabled={disabled}
        id={id}
        name={name}
        ref={inputRef}
        onBlur={onBlur}
        onChange={onChange}
        onClick={onClick}
        onKeyPress={onKeyPress}
        rows={rows}
        value={value}
      />
    </>
  );
}

export default memo(FormBaseTextArea);
