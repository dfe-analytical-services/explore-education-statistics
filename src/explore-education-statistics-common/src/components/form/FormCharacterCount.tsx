import classNames from 'classnames';
import React from 'react';

interface Props {
  id: string;
  maxLength: number;
  value?: string;
}
export default function FormCharacterCount({ id, maxLength, value }: Props) {
  const remaining = maxLength - (value?.trim().length ?? 0);

  return (
    <div
      aria-live="polite"
      className={classNames('govuk-character-count__message', {
        'govuk-hint': remaining >= 0,
        'govuk-error-message': remaining < 0,
      })}
      id={`${id}-info`}
    >
      {remaining >= 0
        ? `You have ${remaining} character${
            remaining !== 1 ? 's' : ''
          } remaining`
        : `You have ${Math.abs(remaining)} character${
            remaining !== -1 ? 's' : ''
          } too many`}
    </div>
  );
}
