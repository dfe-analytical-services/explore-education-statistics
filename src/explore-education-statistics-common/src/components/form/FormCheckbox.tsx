import classNames from 'classnames';
import React, { ChangeEvent, FocusEventHandler, memo, ReactNode } from 'react';

export type OtherCheckboxChangeProps = Pick<FormCheckboxProps, 'label'>;

export type CheckboxChangeEventHandler = (
  event: ChangeEvent<HTMLInputElement>,
  otherCheckboxProps: OtherCheckboxChangeProps,
) => void;

export interface FormCheckboxProps {
  checked?: boolean;
  defaultChecked?: boolean;
  className?: string;
  conditional?: ReactNode;
  id: string;
  hint?: string;
  label: string;
  boldLabel?: boolean;
  name: string;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onChange?: CheckboxChangeEventHandler;
  value: string;
  disabled?: boolean;
}

const FormCheckbox = ({
  checked,
  defaultChecked,
  className,
  conditional,
  id,
  hint,
  label,
  boldLabel = false,
  name,
  onBlur,
  onChange,
  value,
  disabled = false,
}: FormCheckboxProps) => {
  return (
    <>
      <div className={classNames('govuk-checkboxes__item', className)}>
        <input
          aria-describedby={hint ? `${id}-item-hint` : undefined}
          className="govuk-checkboxes__input"
          checked={checked}
          defaultChecked={defaultChecked}
          id={id}
          name={name}
          onBlur={onBlur}
          onChange={event => {
            if (onChange) {
              onChange(event, { label });
            }
          }}
          type="checkbox"
          value={value}
          disabled={disabled}
        />
        <label className="govuk-label govuk-checkboxes__label" htmlFor={id}>
          {boldLabel ? <strong>{label}</strong> : label}
        </label>
        {hint && (
          <span
            id={`${id}-item-hint`}
            className="govuk-hint govuk-checkboxes__hint"
          >
            {hint}
          </span>
        )}
      </div>
      {conditional && (
        <div
          className={classNames('govuk-checkboxes__conditional', {
            'govuk-checkboxes__conditional--hidden': !checked,
          })}
        >
          {conditional}
        </div>
      )}
    </>
  );
};

export default memo(FormCheckbox);
