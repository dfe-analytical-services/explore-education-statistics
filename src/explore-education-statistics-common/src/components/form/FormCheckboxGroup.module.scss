@import '~govuk-frontend/govuk/base';

.selectAll {
  float: left;
  margin-bottom: govuk-spacing(2);
}
