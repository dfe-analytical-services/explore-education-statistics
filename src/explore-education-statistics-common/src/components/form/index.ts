export { default as Form } from './Form';
export { default as FormCheckbox } from './FormCheckbox';
export { default as FormCheckboxGroup } from './FormCheckboxGroup';
export { default as FormCheckboxSearchSubGroups } from './FormCheckboxSearchSubGroups';
export { default as FormCheckboxSearchGroup } from './FormCheckboxSearchGroup';
export { default as FormFieldCheckbox } from './FormFieldCheckbox';
export { default as FormFieldCheckboxGroup } from './FormFieldCheckboxGroup';
export { default as FormFieldCheckboxSearchGroup } from './FormFieldCheckboxSearchGroup';
export { default as FormFieldCheckboxSearchSubGroups } from './FormFieldCheckboxSearchSubGroups';
export { default as FormFieldRadioGroup } from './FormFieldRadioGroup';
export { default as FormFieldSelect } from './FormFieldSelect';
export { default as FormFieldTextArea } from './FormFieldTextArea';
export { default as FormFieldTextInput } from './FormFieldTextInput';
export { default as FormFieldset } from './FormFieldset';
export { default as FormGroup } from './FormGroup';
export { default as FormRadio } from './FormRadio';
export { default as FormRadioGroup } from './FormRadioGroup';
export { default as FormSelect } from './FormSelect';
export { default as FormTextInput } from './FormTextInput';
export { default as FormTextSearchInput } from './FormTextSearchInput';
