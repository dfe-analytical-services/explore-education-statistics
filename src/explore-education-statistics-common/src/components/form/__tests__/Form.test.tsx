import { FormFieldTextInput } from '@common/components/form';
import Yup from '@common/validation/yup';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Field, Formik } from 'formik';
import noop from 'lodash/noop';
import React from 'react';
import Form from '../Form';

describe('Form', () => {
  test('renders error summary from form errors when form is submitted', async () => {
    const { container } = render(
      <Formik
        initialValues={{
          firstName: '',
          lastName: '',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required('First name is required'),
          lastName: Yup.string().required('Last name is required'),
        })}
        onSubmit={() => {}}
      >
        <Form id="test-form">
          The form
          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.getByText('First name is required')).toHaveAttribute(
        'href',
        '#test-form-firstName',
      );
      expect(screen.getByText('Last name is required')).toHaveAttribute(
        'href',
        '#test-form-lastName',
      );
    });

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('does not render errors for fields that do not have errors', async () => {
    const { container } = render(
      <Formik
        initialValues={{
          firstName: '',
          lastName: 'Lastname',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required('First name is required'),
          lastName: Yup.string().required('Last name is required'),
        })}
        onSubmit={() => {}}
      >
        <Form id="test-form">
          The form
          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.queryByText('First name is required')).toBeInTheDocument();
      expect(
        screen.queryByText('Last name is required'),
      ).not.toBeInTheDocument();
    });

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders nested error messages', async () => {
    render(
      <Formik
        initialValues={{
          address: {
            line1: '',
          },
        }}
        validationSchema={Yup.object({
          address: Yup.object({
            line1: Yup.string().required('Line 1 of address is required'),
          }),
        })}
        onSubmit={() => {}}
      >
        <Form id="test-form">
          The form
          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.getByText('Line 1 of address is required')).toHaveAttribute(
        'href',
        '#test-form-addressLine1',
      );
    });
  });

  test('does not render nested error messages for fields that do not have errors', async () => {
    render(
      <Formik
        initialValues={{
          address: {
            line1: 'Line 1',
            line2: '',
          },
        }}
        validationSchema={Yup.object({
          address: Yup.object({
            line1: Yup.string().required('Line 1 of address is required'),
            line2: Yup.string().required('Line 2 of address is required'),
          }),
        })}
        onSubmit={() => {}}
      >
        <Form id="test-form">
          The form
          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(
        screen.queryByText('Line 1 of address is required'),
      ).not.toBeInTheDocument();
      expect(
        screen.queryByText('Line 2 of address is required'),
      ).toBeInTheDocument();
    });
  });

  test('calls `onSubmit` handler when form is submitted successfully', async () => {
    const handleSubmit = jest.fn();

    render(
      <Formik
        initialValues={{
          firstName: 'Firstname',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required(),
        })}
        onSubmit={handleSubmit}
      >
        <Form id="test-form">
          The form
          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalledTimes(1);
    });
  });

  test('renders submit error', async () => {
    const { container } = render(
      <Formik
        initialValues={{
          firstName: 'Firstname',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required(),
        })}
        onSubmit={() => {
          throw new Error('Something went wrong');
        }}
      >
        <Form id="test-form" showSubmitError>
          The form
          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.getByText('Something went wrong')).toHaveAttribute(
        'href',
        '#test-form-submit',
      );
    });

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('removes submit error when form can be submitted successfully', async () => {
    const onSubmit = jest.fn(() => {
      throw new Error('Something went wrong');
    });

    render(
      <Formik
        initialValues={{
          firstName: 'Firstname',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required(),
        })}
        onSubmit={onSubmit}
      >
        <Form id="test-form" showSubmitError>
          The form
          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.queryByText('Something went wrong')).toBeInTheDocument();
    });

    // Stop the onSubmit from throwing error
    onSubmit.mockImplementation();

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(
        screen.queryByText('Something went wrong'),
      ).not.toBeInTheDocument();
    });
  });

  test('removes submit error when form is reset', async () => {
    render(
      <Formik
        initialValues={{
          firstName: 'Firstname',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required(),
        })}
        onSubmit={() => {
          throw new Error('Something went wrong');
        }}
      >
        {formik => (
          <Form id="test-form" showSubmitError>
            The form
            <button type="submit">Submit</button>
            <button type="button" onClick={formik.handleReset}>
              Reset form
            </button>
          </Form>
        )}
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.queryByText('Something went wrong')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('Reset form'));

    await waitFor(() => {
      expect(
        screen.queryByText('Something went wrong'),
      ).not.toBeInTheDocument();
    });
  });

  test('removes submit error when form values are changed', async () => {
    render(
      <Formik
        initialValues={{
          firstName: 'Firstname',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required(),
        })}
        onSubmit={() => {
          throw new Error('Something went wrong');
        }}
      >
        <Form id="test-form" showSubmitError>
          <label htmlFor="firstName">Firstname</label>
          <Field name="firstName" type="text" id="firstName" />

          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.queryByText('Something went wrong')).toBeInTheDocument();
    });

    fireEvent.change(screen.getByLabelText('Firstname'), {
      target: {
        value: 'Another firstname',
      },
    });

    expect(screen.queryByText('Something went wrong')).not.toBeInTheDocument();
  });

  test('focuses error summary on submit', async () => {
    render(
      <Formik
        initialValues={{
          firstName: '',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required(),
        })}
        onSubmit={noop}
      >
        <Form id="test-form" showSubmitError>
          <FormFieldTextInput
            id="test-form-firstName"
            label="First name"
            name="firstName"
          />

          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.getByRole('alert')).toHaveFocus();
    });
  });

  test('does not re-focus error summary when changing input', async () => {
    render(
      <Formik
        initialValues={{
          firstName: '',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required(),
        })}
        onSubmit={noop}
      >
        <Form id="test-form" showSubmitError>
          <FormFieldTextInput
            id="test-form-firstName"
            label="First name"
            name="firstName"
          />

          <button type="submit">Submit</button>
        </Form>
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.getByRole('alert')).toHaveFocus();
    });

    const input = screen.getByLabelText('First name');

    input.focus();

    fireEvent.change(input, {
      target: {
        value: 'a first name',
      },
    });

    expect(screen.getByRole('alert')).not.toHaveFocus();
  });

  test('re-focuses error summary on re-submit', async () => {
    render(
      <Formik
        initialValues={{
          firstName: '',
        }}
        validationSchema={Yup.object({
          firstName: Yup.string().required(),
        })}
        onSubmit={noop}
      >
        {() => (
          <Form id="test-form" showSubmitError>
            <FormFieldTextInput
              id="test-form-firstName"
              label="First name"
              name="firstName"
            />

            <button type="submit">Submit</button>
          </Form>
        )}
      </Formik>,
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.getByRole('alert')).toHaveFocus();
    });

    screen.getByLabelText('First name').focus();
    expect(screen.getByRole('alert')).not.toHaveFocus();

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));

    await waitFor(() => {
      expect(screen.getByRole('alert')).toHaveFocus();
    });
  });
});
