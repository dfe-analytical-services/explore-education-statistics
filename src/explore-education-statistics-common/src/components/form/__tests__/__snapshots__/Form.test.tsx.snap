// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form does not render errors for fields that do not have errors 1`] = `
<form id="test-form">
  <div aria-labelledby="test-form-summary-title"
       class="govuk-error-summary"
       role="alert"
       tabindex="-1"
  >
    <h2 class="govuk-error-summary__title"
        id="test-form-summary-title"
    >
      There is a problem
    </h2>
    <div class="govuk-error-summary__body">
      <ul class="govuk-list govuk-error-summary__list">
        <li>
          <a href="#test-form-firstName">
            First name is required
          </a>
        </li>
      </ul>
    </div>
  </div>
  The form
  <button type="submit">
    Submit
  </button>
</form>
`;

exports[`Form renders error summary from form errors when form is submitted 1`] = `
<form id="test-form">
  <div aria-labelledby="test-form-summary-title"
       class="govuk-error-summary"
       role="alert"
       tabindex="-1"
  >
    <h2 class="govuk-error-summary__title"
        id="test-form-summary-title"
    >
      There is a problem
    </h2>
    <div class="govuk-error-summary__body">
      <ul class="govuk-list govuk-error-summary__list">
        <li>
          <a href="#test-form-firstName">
            First name is required
          </a>
        </li>
        <li>
          <a href="#test-form-lastName">
            Last name is required
          </a>
        </li>
      </ul>
    </div>
  </div>
  The form
  <button type="submit">
    Submit
  </button>
</form>
`;

exports[`Form renders submit error 1`] = `
<form id="test-form">
  <div aria-labelledby="test-form-summary-title"
       class="govuk-error-summary"
       role="alert"
       tabindex="-1"
  >
    <h2 class="govuk-error-summary__title"
        id="test-form-summary-title"
    >
      There is a problem
    </h2>
    <div class="govuk-error-summary__body">
      <ul class="govuk-list govuk-error-summary__list">
        <li>
          <a href="#test-form-submit">
            Something went wrong
          </a>
        </li>
      </ul>
    </div>
  </div>
  The form
  <button type="submit">
    Submit
  </button>
</form>
`;
